api/src/OpenTelemetry/Internal/Logging/Types.hs

```diff
@@ -0,0 +1,162 @@
+  --
+  -- [These short names](https://opentelemetry.io/docs/specs/otel/logs/data-model/#displaying-severity) can be used to represent SeverityNumber in the UI
+  --
+  -- In the contexts where severity participates in less-than / greater-than comparisons SeverityNumber field should be used.
+  -- SeverityNumber can be compared to another SeverityNumber or to numbers in the 1..24 range (or to the corresponding short names).
+  , logRecordBody :: body
+  -- ^ A value containing the body of the log record. Can be for example a human-readable string message (including multi-line) describing the event in a free form or it can be a
+  -- structured data composed of arrays and maps of other values. Body MUST support any type to preserve the semantics of structured logs emitted by the applications.
+  -- Can vary for each occurrence of the event coming from the same source. This field is optional.
+  --
+  -- Type any
+  --    Value of type any can be one of the following:
+  --    - A scalar value: number, string or boolean,
+  --    - A byte array,
+  --    - An array (a list) of any values,
+  --    - A map<string, any>.
+  , logRecordResource :: Maybe MaterializedResources
+  -- ^ Describes the source of the log, aka resource. Multiple occurrences of events coming from the same event source can happen across time and they all have the same value of Resource.
+  -- Can contain for example information about the application that emits the record or about the infrastructure where the application runs. Data formats that represent this data model
+  -- may be designed in a manner that allows the Resource field to be recorded only once per batch of log records that come from the same source. SHOULD follow OpenTelemetry semantic conventions for Resources.
+  -- This field is optional.
+  , logRecordAttributes :: LogAttributes
+  -- ^ Additional information about the specific event occurrence. Unlike the Resource field, which is fixed for a particular source, Attributes can vary for each occurrence of the event coming from the same source.
+  -- Can contain information about the request context (other than Trace Context Fields). The log attribute model MUST support any type, a superset of standard Attribute, to preserve the semantics of structured attributes
+  -- emitted by the applications. This field is optional.
+  }
+  deriving (Functor)
+
+
+{- | Arguments that may be set on LogRecord creation. If observedTimestamp is not set, it will default to the current timestamp.
+If context is not specified it will default to the current context.
+-}
+data LogRecordArguments body = LogRecordArguments
+  { timestamp :: Maybe Timestamp
+  , observedTimestamp :: Maybe Timestamp
+  , context :: Maybe Context
+  , severityText :: Maybe Text
+  , severityNumber :: Maybe Int
+  , body :: body
+  , attributes :: LogAttributes
+  }
+
+
+data SeverityNumber = SeverityNumber {shortName :: Maybe Text, severityInt :: !Int64} deriving (Read, Show)
```

Comment:
Several implementations are possible here, but this one is simple to use because `mkSeverityNumber`, `shortName`, and `severityInt` are the most common if not the only operations needed out of a severity number.